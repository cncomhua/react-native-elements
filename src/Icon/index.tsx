import { withTheme } from '../config';
import { Icon, IconProps, IconNode, IconObject, IconType } from './Icon';

export { Icon };
export type { IconProps, IconNode, IconObject, IconType };
/**
 * RNE Icon
 *
 * **Preview**
 *
 * ![URL](URL)
 *
 * [**Documentation**](https://reactnativeelements.com/docs/icon)
 */
export default withTheme(Icon, 'Icon');
