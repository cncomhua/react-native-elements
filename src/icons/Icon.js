import React from 'react';
import PropTypes from 'prop-types';
import { TouchableHighlight, View, Text as NativeText } from 'react-native';
import getIconType from '../helpers/getIconType';
import ViewPropTypes from '../config/ViewPropTypes';

class Icon extends React.PureComponent {
  render() {
    const {
      type,
      name,
      size,
      color,
      iconStyle,
      component,
      underlayColor,
      reverse,
      raised,
      containerStyle,
      reverseColor,
      onPress,
      ...attributes
    } = this.props;

    let Component = View;
    if (onPress) {
      Component = TouchableHighlight;
    }
    if (component) {
      Component = component;
    }
    let Icon;
    if (!type) {
      Icon = getIconType('material');
    } else {
      Icon = getIconType(type);
    }

    const styles = this.context.theme.icon;

    return (
      <Component
        underlayColor={reverse ? color : underlayColor || color}
        style={[
          (reverse || raised) && styles.button,
          (reverse || raised) && {
            borderRadius: size + 4,
            height: size * 2 + 4,
            width: size * 2 + 4,
          },
          raised && styles.raised,
          {
            backgroundColor: reverse ? color : raised ? 'white' : 'transparent',
            alignItems: 'center',
            justifyContent: 'center',
          },
          containerStyle && containerStyle,
        ]}
        onPress={onPress}
        {...attributes}
      >
        <Icon
          style={[{ backgroundColor: 'transparent' }, iconStyle && iconStyle]}
          size={size}
          name={name}
          color={reverse ? reverseColor : color}
        />
      </Component>
    );
  }
}

Icon.propTypes = {
  type: PropTypes.string,
  name: PropTypes.string,
  size: PropTypes.number,
  color: PropTypes.string,
  component: PropTypes.func,
  underlayColor: PropTypes.string,
  reverse: PropTypes.bool,
  raised: PropTypes.bool,
  containerStyle: ViewPropTypes.style,
  iconStyle: NativeText.propTypes.style,
  onPress: PropTypes.func,
  reverseColor: PropTypes.string,
};

Icon.defaultProps = {
  underlayColor: 'white',
  reverse: false,
  raised: false,
  size: 24,
  color: 'black',
  reverseColor: 'white',
};

Icon.contextTypes = {
  theme: PropTypes.object.isRequired,
};

export default Icon;
