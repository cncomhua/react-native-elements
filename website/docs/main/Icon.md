---


id: icon


title: Icon


slug: /icon

---



import Usage from './usage/Icon/Icon.md'



Icons are visual indicators usually used to describe action or intent.

They are also used for displaying information.



## Usage


<Usage />

---


## Props

### Icon




## Reference

### Icon

