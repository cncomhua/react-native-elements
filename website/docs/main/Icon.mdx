---
id: icon

title: Icon

slug: /icon
---

import Usage from './usage/Icon/Icon.mdx';

import Props from './props/Icon.mdx';

import { Icon } from 'react-native-elements';

import Tabs from '@theme/Tabs';

import TabItem from '@theme/TabItem';

Icons are visual indicators usually used to describe action or intent.

They are also used for displaying information.

## Usage

<Usage />

## Props

<Props />
