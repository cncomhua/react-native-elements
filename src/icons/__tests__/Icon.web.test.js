import React from 'react';
import {
  Text,
  TouchableHighlight,
  TouchableWithoutFeedback,
  View,
  Platform,
} from 'react-native';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import parseSamples, {
  assignRef,
  genRefId,
} from 'enzyme-styleguidist-sample-parser';
import options, {
  snapShot,
  buildJsxForGuideMethod,
  ensureCalled,
  onlyEnsureCalled,
  onlySnapshots,
} from '../../../samples';
import Icon from '../Icon';

const props = {
  noProps: {
    component: Icon,
    enzyme: {
      tests: {
        shallow: { snapshot: snapShot() },
      },
    },
    styleguidist: {},
  },
  name: {
    component: Icon,
    props: { name: 'pets' },
    enzyme: {
      tests: { shallow: { snapshot: snapShot() } },
    },
    styleguidist: {},
  },
  type: {
    component: Icon,
    props: { name: 'sms', type: 'octicons' },
    enzyme: {
      tests: { shallow: { snapshot: snapShot() } },
    },
  },
  size: {
    component: Icon,
    props: { name: 'pets', size: 50 },
    enzyme: {
      tests: { shallow: { snapshot: snapShot() } },
    },
  },
  color: {
    component: Icon,
    props: { name: 'pets', color: '#071' },
    enzyme: {
      tests: { shallow: { snapshot: snapShot() } },
    },
  },
  iconStyle: {
    component: Icon,
    props: { name: 'pets', iconStyle: { transform: [{ rotateZ: '-45deg' }] } },
    enzyme: {
      tests: { shallow: { snapshot: snapShot() } },
    },
  },
  component: {
    component: Icon,
    props: { name: 'pets', component: TouchableWithoutFeedback },
    enzyme: {
      tests: { shallow: { snapshot: snapShot() } },
    },
    styleguidist: {
      getJsxString: () => {
        return `  const TouchableWithoutFeedback = RN.TouchableWithoutFeedback;
  <Icon
  name='pets'
    component={TouchableWithoutFeedback}
  />`;
      },
    },
  },
  onPress: {
    component: Icon,
    props: {
      name: 'pets',
      onPress: () => {
        console.log('pressed');
      },
    },
    enzyme: {
      tests: { shallow: { snapshot: snapShot() } },
    },
  },
  onLongPress: {
    component: Icon,
    props: {
      name: 'pets',
      onLongPress: () => {
        console.log('pressed, long');
      },
    },
    enzyme: {
      tests: { shallow: { snapshot: snapShot() } },
    },
  },
  underlayColor: {
    component: Icon,
    props: {
      name: 'pets',
      onPress: () => {
        console.log('pressed');
      },
      underlayColor: 'yellow',
    },
    enzyme: {
      tests: { shallow: { snapshot: snapShot() } },
    },
  },
  reverse: {
    component: Icon,
    props: { name: 'pets', reverse: true },
    enzyme: {
      tests: { shallow: { snapshot: snapShot() } },
    },
  },
  raised: {
    component: Icon,
    props: { name: 'pets', raised: true },
    enzyme: {
      tests: { shallow: { snapshot: snapShot() } },
    },
  },
  containerStyle: {
    component: Icon,
    props: { name: 'pets', containerStyle: { padding: 50 } },
    enzyme: {
      tests: { shallow: { snapshot: snapShot() } },
    },
  },
  reverseColor: {
    component: Icon,
    props: { name: 'pets', reverse: true, reverseColor: '#071' },
    enzyme: {
      tests: { shallow: { snapshot: snapShot() } },
    },
  },
};
const samples = {
  props,
};

parseSamples(
  {
    Icons: { sectionComponents: { Icon: { samples } } },
  },
  options
);

describe('Icon component', () => {
  it('should render without issues', () => {
    const component = shallow(<Icon name="wifi" />);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render with icon type', () => {
    const component = shallow(
      <Icon
        name="alert"
        type="octicon"
        reverse
        color="red"
        iconStyle={{ backgroundColor: 'peru' }}
      />
    );

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should have onPress event', () => {
    const onPress = jest.fn();
    const component = shallow(<Icon onPress={onPress} name="wifi" />);

    component.simulate('press');
    expect(onPress).toHaveBeenCalledTimes(1);
  });
});
