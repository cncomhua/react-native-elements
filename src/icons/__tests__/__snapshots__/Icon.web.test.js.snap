// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon component should render with icon type 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "red",
      "borderRadius": 28,
      "height": 52,
      "justifyContent": "center",
      "margin": 7,
      "width": 52,
    }
  }
>
  <Icon
    color="white"
    name="alert"
    size={24}
    style={
      Object {
        "backgroundColor": "peru",
      }
    }
  />
</View>
`;

exports[`Icon component should render without issues 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "justifyContent": "center",
    }
  }
>
  <Icon
    color="black"
    name="wifi"
    size={24}
    style={
      Object {
        "backgroundColor": "transparent",
      }
    }
  />
</View>
`;

exports[`Icons: Icon: Props Icons: Icon: [ prop: color ] enzyme-shallow: snapshot: Icons: Icon: [ prop: color ] enzyme-shallow: snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "justifyContent": "center",
    }
  }
>
  <Icon
    color="#071"
    name="pets"
    size={24}
    style={
      Object {
        "backgroundColor": "transparent",
      }
    }
  />
</View>
`;

exports[`Icons: Icon: Props Icons: Icon: [ prop: component ] enzyme-shallow: snapshot: Icons: Icon: [ prop: component ] enzyme-shallow: snapshot 1`] = `
<TouchableWithoutFeedback
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "justifyContent": "center",
    }
  }
>
  <Icon
    color="black"
    name="pets"
    size={24}
    style={
      Object {
        "backgroundColor": "transparent",
      }
    }
  />
</TouchableWithoutFeedback>
`;

exports[`Icons: Icon: Props Icons: Icon: [ prop: containerStyle ] enzyme-shallow: snapshot: Icons: Icon: [ prop: containerStyle ] enzyme-shallow: snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "justifyContent": "center",
      "padding": 50,
    }
  }
>
  <Icon
    color="black"
    name="pets"
    size={24}
    style={
      Object {
        "backgroundColor": "transparent",
      }
    }
  />
</View>
`;

exports[`Icons: Icon: Props Icons: Icon: [ prop: iconStyle ] enzyme-shallow: snapshot: Icons: Icon: [ prop: iconStyle ] enzyme-shallow: snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "justifyContent": "center",
    }
  }
>
  <Icon
    color="black"
    name="pets"
    size={24}
    style={
      Object {
        "backgroundColor": "transparent",
        "transform": Array [
          Object {
            "rotateZ": "-45deg",
          },
        ],
      }
    }
  />
</View>
`;

exports[`Icons: Icon: Props Icons: Icon: [ prop: name ] enzyme-shallow: snapshot: Icons: Icon: [ prop: name ] enzyme-shallow: snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "justifyContent": "center",
    }
  }
>
  <Icon
    color="black"
    name="pets"
    size={24}
    style={
      Object {
        "backgroundColor": "transparent",
      }
    }
  />
</View>
`;

exports[`Icons: Icon: Props Icons: Icon: [ prop: noProps ] enzyme-shallow: snapshot: Icons: Icon: [ prop: noProps ] enzyme-shallow: snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "justifyContent": "center",
    }
  }
>
  <Icon
    color="black"
    size={24}
    style={
      Object {
        "backgroundColor": "transparent",
      }
    }
  />
</View>
`;

exports[`Icons: Icon: Props Icons: Icon: [ prop: onLongPress ] enzyme-shallow: snapshot: Icons: Icon: [ prop: onLongPress ] enzyme-shallow: snapshot 1`] = `
<TouchableHighlight
  activeOpacity={0.85}
  onLongPress={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "justifyContent": "center",
    }
  }
  underlayColor="white"
>
  <Icon
    color="black"
    name="pets"
    size={24}
    style={
      Object {
        "backgroundColor": "transparent",
      }
    }
  />
</TouchableHighlight>
`;

exports[`Icons: Icon: Props Icons: Icon: [ prop: onPress ] enzyme-shallow: snapshot: Icons: Icon: [ prop: onPress ] enzyme-shallow: snapshot 1`] = `
<TouchableHighlight
  activeOpacity={0.85}
  onPress={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "justifyContent": "center",
    }
  }
  underlayColor="white"
>
  <Icon
    color="black"
    name="pets"
    size={24}
    style={
      Object {
        "backgroundColor": "transparent",
      }
    }
  />
</TouchableHighlight>
`;

exports[`Icons: Icon: Props Icons: Icon: [ prop: raised ] enzyme-shallow: snapshot: Icons: Icon: [ prop: raised ] enzyme-shallow: snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "borderRadius": 28,
      "boxShadow": "
        0 3px 4px 0px rgba(0, 0, 0, 0.14),
        0 3px 3px -2px rgba(0, 0, 0, 0.2)
      ",
      "height": 52,
      "justifyContent": "center",
      "margin": 7,
      "width": 52,
    }
  }
>
  <Icon
    color="black"
    name="pets"
    size={24}
    style={
      Object {
        "backgroundColor": "transparent",
      }
    }
  />
</View>
`;

exports[`Icons: Icon: Props Icons: Icon: [ prop: reverse ] enzyme-shallow: snapshot: Icons: Icon: [ prop: reverse ] enzyme-shallow: snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "black",
      "borderRadius": 28,
      "height": 52,
      "justifyContent": "center",
      "margin": 7,
      "width": 52,
    }
  }
>
  <Icon
    color="white"
    name="pets"
    size={24}
    style={
      Object {
        "backgroundColor": "transparent",
      }
    }
  />
</View>
`;

exports[`Icons: Icon: Props Icons: Icon: [ prop: reverseColor ] enzyme-shallow: snapshot: Icons: Icon: [ prop: reverseColor ] enzyme-shallow: snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "black",
      "borderRadius": 28,
      "height": 52,
      "justifyContent": "center",
      "margin": 7,
      "width": 52,
    }
  }
>
  <Icon
    color="#071"
    name="pets"
    size={24}
    style={
      Object {
        "backgroundColor": "transparent",
      }
    }
  />
</View>
`;

exports[`Icons: Icon: Props Icons: Icon: [ prop: size ] enzyme-shallow: snapshot: Icons: Icon: [ prop: size ] enzyme-shallow: snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "justifyContent": "center",
    }
  }
>
  <Icon
    color="black"
    name="pets"
    size={50}
    style={
      Object {
        "backgroundColor": "transparent",
      }
    }
  />
</View>
`;

exports[`Icons: Icon: Props Icons: Icon: [ prop: type ] enzyme-shallow: snapshot: Icons: Icon: [ prop: type ] enzyme-shallow: snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "justifyContent": "center",
    }
  }
>
  <Icon
    color="black"
    name="sms"
    size={24}
    style={
      Object {
        "backgroundColor": "transparent",
      }
    }
  />
</View>
`;

exports[`Icons: Icon: Props Icons: Icon: [ prop: underlayColor ] enzyme-shallow: snapshot: Icons: Icon: [ prop: underlayColor ] enzyme-shallow: snapshot 1`] = `
<TouchableHighlight
  activeOpacity={0.85}
  onPress={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "justifyContent": "center",
    }
  }
  underlayColor="yellow"
>
  <Icon
    color="black"
    name="pets"
    size={24}
    style={
      Object {
        "backgroundColor": "transparent",
      }
    }
  />
</TouchableHighlight>
`;
